### PR TITLE
Write mode

### PR DESCRIPTION
**Pull request by wafcio**

### CODE REVIEW

File: lib/rom/csv/dataset/write_operations.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        def update(index, tuple)
+          data[index].merge!(tuple)
+        end
+
+        def delete(tuple)
+          data.delete(tuple)
+        end
+
+        def write
+          ::CSV.open(path, 'wb') do |csv|
```
Comment (splattael) [inline lib/rom/csv/dataset/write_operations.rb]: We should pass `options` passed via `ROM.setup(:csv, "file", options)` to `CSV.open`.

**Comment by splattael** [comment]
I like it! :+1: 

I am not sure about `LoadData` module. It could be inlined into `Dataset` as `reload` and `load_data_from_file` are used privately. Otherwise the code is hard to follow.

File: lib/rom/csv/dataset/write_operations.rb (hunk 1 above)
Comment (gotar) [inline lib/rom/csv/dataset/write_operations.rb]: It was before done for `CSV.table ...`

**Comment by solnic** [comment]
Would be nice to reuse the logic from `Memory::Dataset`. If inheritance no longer works we could extract it into a shared module.

File: lib/rom/csv/commands/create.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
+require 'rom/commands'
+require 'rom/commands/create'
+
+module ROM
+  module CSV
+    module Commands
+      class Create < ROM::Commands::Create
+        def execute(tuples)
+          insert_tuples = Array([tuples]).flatten.map do |tuple|
```
Comment (wijet) [inline lib/rom/csv/commands/create.rb]: I believe `Array([tuples]).flatten` is not needed, `Array(tuples).flatten` does the same. Moreover this flatten can be removed as well, but not sure about that.

Comment (AMHOL) [inline lib/rom/csv/commands/create.rb]: Was thinking the same, would go with:

``` ruby
Array(tuples).flat_map do |tuple|
```

**Comment by solnic** [comment]
Awesome work :) I've got a couple of general comments:
- we need to know the header - I'd say if you want to write to csv files then you should provide the header upfront and the adapter won't have to try to be smart and infer it from the first tuple. It would also eliminate the need to mutate dataset via `add_header`
- since it adds an overhead of reloading data when re-reading (ie see `to_a`) then I'd say we should have 2 types of dataset - read and read/write - this could be done in a separate PR though, I'm OK with merging this one in first to avoid a long-lasting PR and risking a burnout :)
- I'd double check if maybe there's a way to reduce mutability of the dataset to the minimum

File: lib/rom/csv/commands/create.rb (hunk 2 above)
Comment (wafcio) [inline lib/rom/csv/commands/create.rb]: ok I will remove it, but the same code exist in rom-sql

Comment (wafcio) [inline lib/rom/csv/commands/create.rb]: This code come from rom-sql, I will leave [tuples].flatten.map

Array(tuples) doesn't work the same like [tuples] and flatten.map doesn't work the same like flat_map

**Comment by wafcio** [comment]
@solnic Memory::Dataset doesn't work in this case. I created new dataset because:
- remember path
- way to reload data

some of Memory::Dataset methods like find_all, have magic word "return" which return filtered set instead of dataset wrapper.

**Comment by wafcio** [comment]
removed add_header, but I am not sure that we shouldn't have way to modify headers. When csv headers are different than our class attributes.

**Comment by wafcio** [comment]
if you have more suggestion what can be changed in this code let me know.

File: lib/rom/csv/dataset/read_operations.rb [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+          data.select! { |d| yield(d) }
+        end
+
+        def map!
+          data.map! { |d| yield(d) }
+        end
+
+        def sort_by!
+          data.sort_by! { |d| yield(d) }
+        end
```
Comment (solnic) [inline lib/rom/csv/dataset/read_operations.rb]: why did you use bang-methods?

Comment (wafcio) [inline lib/rom/csv/dataset/read_operations.rb]: I used ! because it works in place and uses less memory (in theory).

Comment (solnic) [inline lib/rom/csv/dataset/read_operations.rb]: This will mutate the dataset so it will affect relations. In general relations/datasets shouldn't be mutable wrt queries. Only insert/update/delete can mutate relation/dataset.

Comment (wafcio) [inline lib/rom/csv/dataset/read_operations.rb]: I can change it to immutable if you want, but I am not sure that immutable dataset is good. When file is very big, in a worst-case scenario it will take n \* original dataset size.

Comment (wafcio) [inline lib/rom/csv/dataset/read_operations.rb]: immutable dataset will not be working with dalete command.

**Comment by wafcio** [comment]
I have just changed <pre>data.select! ...</pre> to <pre>@data = data.select ...</pre> Is it ok or not ?

**Comment by solnic** [comment]
Let me tweak stuff in ROM so that this adapter can leverage existing features. There's too much duplication between the new dataset and memory dataset. I'm also not sure if it's a good idea to make dataset responsible for dealing with the source file. Maybe it would be better to move this to commands. I'll get back to you as soon as stuff on ROM side is ready. Hopefully tomorrow.

File: lib/rom/csv/dataset/read_operations.rb [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+          @data = data.select { |d| yield(d) }
+        end
+
+        def map
+          @data = data.map { |d| yield(d) }
+        end
+
+        def sort_by
+          @data = data.sort_by { |d| yield(d) }
+        end
```
Comment (solnic) [inline lib/rom/csv/dataset/read_operations.rb]: I'll make this redundant and we'll be able to remove it completely but for the sake of explanation - this is not OK because query methods mutates the dataset. Consider this:

``` ruby
users = rom.relation(:users) # all users
users.restrict(name: 'Jane')

# now we no longer have all users since `restrict` mutated the dataset so:
users.to_a # this returns restricted users, which is a severe bug
```

I hope this explains it :)

**Comment by wafcio** [comment]
Ok

**Comment by solnic** [comment]
OK you should be able to subclass `ROM::Memory::Dataset` and add `:path` option now:

``` ruby
class Dataset < ROM::Memory::Dataset
  option :path, reader: true

  def reload!
    @data = load_data
  end

  def sync!
    write_data and reload!
  end
end
```

I hope this helps

**Comment by solnic** [comment]
Just a quick update that I haven't forgotten about this one. It's just very low (unfortunately) on my list and I've been too busy with other rom-\* stuff. Also, I believe it's a non-trivial thing to solve at the end of the day. I'll do my best to go through this PR ASAP and merge it in.

**Comment by solnic** [comment]
@wafcio starting from ROM 0.8.0 all commands have access to source relation, this means if you do `rom.command(:users).update.by_id(1)` then this command will know both its restricted relation and the source one (which is the whole users relation). It should allow adapters like csv to update things properly.

**Comment by wafcio** [comment]
ok, I must take breath from coding after hours. I will probably back to this PR on Thursday or weekend.

**Comment by wafcio** [comment]
@solnic updated

**Comment by solnic** [comment]
It's merged :)
